core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -157,22 +164,32 @@ private class ShuffleStatus(
       invalidateSerializedMapOutputStatusCache()
     }
     mapStatuses(mapIndex) = status
+    mapIdToMapIndex(status.mapId) = mapIndex
+  }
+
+  /**
+   * Get the map output that corresponding to a given mapId.
+   */
+  def getMapStatus(mapId: Long): Option[MapStatus] = withReadLock {
+    mapIdToMapIndex.get(mapId).map(mapStatuses(_))
   }
 
   /**
    * Update the map output location (e.g. during migration).
    */
   def updateMapOutput(mapId: Long, bmAddress: BlockManagerId): Unit = withWriteLock {
     try {
-      val mapStatusOpt = mapStatuses.find(x => x != null && x.mapId == mapId)
+      val mapIndex = mapIdToMapIndex.get(mapId)
+      val mapStatusOpt = Option(mapIndex.map(mapStatuses(_)).getOrElse(null))
       mapStatusOpt match {
         case Some(mapStatus) =>
           logInfo(s"Updating map output for ${mapId} to ${bmAddress}")
           mapStatus.updateLocation(bmAddress)
           invalidateSerializedMapOutputStatusCache()
         case None =>
-          val index = mapStatusesDeleted.indexWhere(x => x != null && x.mapId == mapId)
-          if (index >= 0 && mapStatuses(index) == null) {
+          if (mapIndex.nonEmpty && mapStatuses(mapIndex.get) == null &&
+              mapStatusesDeleted(mapIndex.get).mapId == mapId) {
```

Comment:
We dont need to check for `mapStatuses(mapIndex.get) == null` - it should already be `null` since we are in the `None` case.

```suggestion
          if (mapIndex.map(mapStatusesDeleted).exists(_.mapId == mapId) {
```